### DISCUSSION TITLE
Custom Placement of Two-Row Tall Cells

### DISCUSSION THREAD

**Discussion by harshkhandeparkar** [The-OpenROAD-Project/OpenROAD#4442] [Q&A] [answered]
In the OpenFASoC [LDO generator](https://github.com/idea-fasoc/OpenFASOC/tree/main/openfasoc/generators/ldo-gen), a two-row tall comparator latch is custom placed using the `place_cell` command in a [tcl script](https://github.com/idea-fasoc/OpenFASOC/blob/f7b3440c485ff2accbf549aa11a12a75c01186c1/openfasoc/generators/ldo-gen/flow/scripts/openfasoc/custom_place.tcl#L7).

This used to work in earlier versions of OpenROAD, but recently the CI is [failing](https://github.com/idea-fasoc/OpenFASOC/actions/runs/7301239627/job/19897490889#step:4:1990) with the following error:
```
[ERROR DPL-0017] cannot place instance cmp1.
```
 The custom placement is done before running the `detailed_placement` command in the [detail_place.tcl](https://github.com/idea-fasoc/OpenFASOC/blob/main/openfasoc/generators/ldo-gen/flow/scripts/detail_place.tcl) script. If the `detailed_placement` command is removed, the following error is returned:
```
[ERROR DPL-0044] Cell cmp1 with height 8140 is taller than any row.
Error: detail_place.tcl, 26 DPL-0044
```

Relevant Information:
- OpenROAD Version hash: `e89829335596b351ce665dcc3b73619c0b191c14`
- Comparator Latch Cell: [LEF](https://github.com/idea-fasoc/OpenFASOC/blob/main/openfasoc/generators/ldo-gen/blocks/sky130hvl/lef/LDO_COMPARATOR_LATCH.lef) / [GDS](https://github.com/idea-fasoc/OpenFASOC/blob/main/openfasoc/generators/ldo-gen/blocks/sky130hvl/gds/LDO_COMPARATOR_LATCH.gds)
- Custom placement script: [custom_place.tcl](https://github.com/idea-fasoc/OpenFASOC/blob/main/openfasoc/generators/ldo-gen/flow/scripts/openfasoc/custom_place.tcl)

Is there a way to place this cell (not as a macro) without placement violations?

cc: @msaligane

**Comment by rovinski** [comment]
Have you tried marking the cells as `FIXED` instead of `PLACED`?

**Reply by harshkhandeparkar** [reply]
`FIXED` gives the following error.
```
[ERROR PAD-9188] Invalid placement status FIXED, must be one of either PLACED or FIRM.
```

If you meant `FIRM`, then yes, I have tried it, and it gives the same `taller than any row` error even with `detailed_placement` on.

**Comment by maliberty** [comment]
We have recently being doing work related to multi-height cells.  The error indicates that the detailed placer is trying to legalize this instance which is sounds like you don't want.  FIRM should work for that purpose.  Can you provide a test case?

**Reply by harshkhandeparkar** [reply]
We tried with `FIRM`. It produces the `Cell cmp1 with height 8140 is taller than any row.` error with or without `detailed_placement` enabled.

How should I provide a test case? Would the odb output of the previous step work?

**Comment by maliberty** [comment]
Your other option is to provide double height site rows.

**Reply by harshkhandeparkar** [reply]
How can that be done?

**Reply by maliberty** [reply, marked as answer]
You would need a corresponding site.  You would set PLACE_SITE to a list of your regular and double height site.

**Reply by harshkhandeparkar** [reply]
Using the `unithvdbl` site in the LEF worked. Thanks!
